File: packages_py/runtime_template_resolver/src/runtime_template_resolver/extractor.py
from typing import List, Dict, Any
from .patterns import PATTERNS

def extract_placeholders(template: str) -> List[Dict[str, Any]]:
    """Extract all placeholders from a template string."""
    placeholders = []
    
    # Extract mustache placeholders
    for match in PATTERNS["MUSTACHE"].finditer(template):
        placeholders.append({
            "raw": match.group(0),
            "path": match.group(1).strip(),
            "default": match.group(2),
            "start": match.start(),
            "end": match.end(),
            "syntax": "MUSTACHE"
        })
        
    # Extract dot path placeholders
    for match in PATTERNS["DOT_PATH"].finditer(template):
        placeholders.append({
            "raw": match.group(0),
            "path": match.group(1),
            "default": None,
            "start": match.start(),
            "end": match.end(),
            "syntax": "DOT_PATH"
        })
        
    return placeholders

File: packages_mjs/runtime-template-resolver/src/extractor.ts
import { PATTERNS } from './patterns.js';

export interface Placeholder {
    raw: string;
    path: string;
    default?: string;
    start: number;
    end: number;
    syntax: 'MUSTACHE' | 'DOT_PATH';
}

export function extractPlaceholders(template: string): Placeholder[] {
    const placeholders: Placeholder[] = [];

    // Reset lastIndex for global regex
    PATTERNS.MUSTACHE.lastIndex = 0;
    PATTERNS.DOT_PATH.lastIndex = 0;

    let match;
    while ((match = PATTERNS.MUSTACHE.exec(template)) !== null) {
        placeholders.push({
            raw: match[0],
            path: match[1]?.trim(),
            default: match[2],
            start: match.index,
            end: match.index + match[0].length,
            syntax: 'MUSTACHE'
        });
    }

    while ((match = PATTERNS.DOT_PATH.exec(template)) !== null) {
        placeholders.push({
            raw: match[0],
            path: match[1],
            default: undefined,
            start: match.index,
            end: match.index + match[0].length,
            syntax: 'DOT_PATH'
        });
    }

    return placeholders;
}
